wemake_python_styleguide/visitors/tokenize/statements.py

```diff
@@ -256,3 +258,111 @@ def _post_visit(self) -> None:
+        self._is_comprehension:
+        Flag is set if current clause is identified as a list comprehension.
+        self._seen_clause_in_line:
+        Flag is set when the current line already contains a clause, which
+        is either the action, each for loop, or the conditional. Starts off
+        as True to account for the action, which we don't actually visit.
+        self._seen_nl:
+        Flag for if we've seen any logical newlines, indicating this is a
+        multiline comprehension
+        self._potential_violation = False
+        Flag for when we see multiple clauses in one line. Only a violation
+        if this is a multiline comprehension
+        self._reported:
+        Flag tracks whether we've already reported this violation.
+        """
+        self.is_comprehension = False
+        self.seen_clause_in_line = False
+        self.seen_nl = False
+        self.potential_violation = False
+        self.reported = False
+
+
+@final
+@alias('visit_any_left_bracket', (
+    'visit_lsqb',
+    'visit_lbrace',
+))
+@alias('visit_any_right_bracket', (
+    'visit_rsqb',
+    'visit_rbrace',
+))
+class InconsistentComprehensionVisitor(BaseTokenVisitor):
+    """
+    Visitor for checking inconsistent comprehension syntax.
+
+    Checks if comprehensions either use only one line or inserts a newline
+    for each clause (i.e. bracket, action, for loop(s), and conditional)
+    """
+
+    def __init__(self, *args, **kwargs) -> None:
+        """
+        Initializes stack of bracket contexts.
+
+        Creates an empty stack for bracket contexts to accomodate for nested
+        comprehensions.
+        """
+        super().__init__(*args, **kwargs)
+        self._bracket_stack: List[InconsistentComprehensionContext] = []
+        self._ctxt: Optional[InconsistentComprehensionContext] = None
+
+    def visit_any_left_bracket(self, token: tokenize.TokenInfo) -> None:
+        """Sets self._inside_brackets to True if left bracket found."""
+        self._bracket_stack.append(InconsistentComprehensionContext())
+        self._ctxt = self._bracket_stack[-1]
+
+    def visit_any_right_bracket(self, token: tokenize.TokenInfo) -> None:
+        """Resets environment if right bracket is encountered."""
+        self._bracket_stack.pop()
+        self._ctxt = self._bracket_stack[-1] if self._bracket_stack else None
+
+    def visit_nl(self, token: tokenize.TokenInfo) -> None:
+        """Sets appropriate flags to True if nl encountered inside brackets."""
+        if self._ctxt:
+            self._ctxt.seen_nl = True
+            self._ctxt.seen_clause_in_line = False
+            self._check_violation(token)
+
+    def visit_name(self, token: tokenize.TokenInfo) -> None:
+        """
+        Sets flags for comprehension and potential violation as appropriate.
+
+        Conditionally dependent on the flag signifying inside brackets.
+        """
+        if self._ctxt:
+            if token.string in {'for', 'if'}:
```

Comment:
`{'for', 'if'}` should be a class-level `frozenset` constant.

Reply:
Logic here was changed, making this no longer necessary.